src/Skia/Avalonia.Skia/FormattedTextImpl.cs

```diff
@@ -142,21 +142,16 @@ public Rect HitTestTextPosition(int index)
         {
             var rects = GetRects();
 
-            if (index < 0 || index >= rects.Count)
-            {
-                var r = rects.LastOrDefault();
-                return new Rect(r.X + r.Width, r.Y, 0, _lineHeight);
-            }
-
             if (rects.Count == 0)
             {
-                return new Rect(0, 0, 1, _lineHeight);
+                var x = TransformX(0, 0, _paint.TextAlign);
+                return new Rect(x, 0, 0, _lineHeight);
             }
 
-            if (index == rects.Count)
+            if (index < 0 || index >= rects.Count)
```

Comment:
I would rewrite the code a bit in the following way:
```
if (string.IsNullOrEmpty(Text))
{
    //align carret for empty text
}

if (index < 0)
{
    // align carret at the first position
}

if(index >= Text.Length)
{
    //align carret at the last position
}

var rects = GetRects();

//align carret at glyph position
```

Reply:
After a bit thinking: Not sure I understand what the difference between first two cases.
I thought caret position for empty text is the same as for the first position?

Also I'm not sure how to apply `textBounds.Width` for last position case. Are you certain it will work?

Reply:
Empty text has a width of zero. But the TextBox itself takes all space it gets. For non empty text the sum of all glyphs makes up the width. Then you arrange the glyphs in combination with the text alignment. If the index is below the first position the caret should be at the left side of the text. If the index is above the length of character the caret should be at the right side.

Left side - > (0,LineHeight) + AlignmentOffset
Right side -> (width, LineHeight) + AlignmentOffset

Reply:
Thank you. I've updated the code. Does it look better now? Asking because to me it's not :-)

Reply:
That looks great 👍